### PR TITLE
feat(dashy): Enable XYAxis range selection

### PR DESCRIPTION
**Pull request by GordonSmith**
image

image

### CODE REVIEW

**Comment by GordonSmith** [comment]
@jbrundage please review

File: packages/chart/src/XYAxis.ts [hunk 1]
```diff
@@ -194,7 +194,7 @@ export class XYAxis extends SVGWidget {
         const currSel: any = d3BrushSelection(this.svgBrush.node());
         if (currSel) {
             selected = this.data().filter(function (d) {
-                const pos = context.dataPos(d[0]) + context.domainAxis.bandwidth() / 2;
```
Comment (jbrundage) [inline packages/chart/src/XYAxis.ts]: Why was ` + context.domainAxis.bandwidth() / 2` removed? Should the same thing happen to ` + context.valueAxis.bandwidth() / 2` on line 199?

Comment (GordonSmith) [reply packages/chart/src/XYAxis.ts]: It was not needed (once I started highlighting which ones were selected it was pretty obvious).  and yes probably...

Comment (GordonSmith) [reply packages/chart/src/XYAxis.ts]: Which may mean that when the X-Axis is linear it will be needed

Comment (GordonSmith) [reply packages/chart/src/XYAxis.ts]: Just tested seems to work as expected - the 2D selection is still a bit wonky though (the handles don't move with it and there are no handles on the vertical...

File: packages/marshaller/src/ddl2/activities/filter.ts [hunk 2]
```diff
@@ -100,16 +100,27 @@ export class ColumnMapping extends PropertyExt {
...
-        if (isString) {
-            fs = fs.trim();
+        let fs = filterSelection.length ? filterSelection.map(sel => {
+            let retVal = sel[rf];
+            const isString = typeof retVal === "string";
+            if (isString) {
+                retVal = retVal.trim();
+            }
+            return retVal;
+        }).join(", ") : undefined;
```
Comment (jbrundage) [inline packages/marshaller/src/ddl2/activities/filter.ts]: The `.join(", ")` will cause everything to return as a string (unless it's undefined)

Comment (jbrundage) [reply packages/marshaller/src/ddl2/activities/filter.ts]: ...which will cause issues in the switch statement within `createFilter`

File: packages/marshaller/src/ddl2/model/element.ts [hunk 3]
```diff
@@ -79,8 +79,13 @@ export class Element extends PropertyExt {
...
         this._elementContainer = ec;
-        vizID++;
-        this._id = `e_${vizID}`;
+        while (true) {
+            vizID++;
+            this._id = `e_${vizID}`;
+            if (!this._elementContainer.elementExists(this._id)) {
+                break;
+            }
+        }
```
Comment (jbrundage) [inline packages/marshaller/src/ddl2/model/element.ts]: What could cause duplicate `this._id`? It seems like this should never be necessary?

File: packages/marshaller/src/ddl2/activities/filter.ts (hunk 2 above)
Comment (GordonSmith) [reply packages/marshaller/src/ddl2/activities/filter.ts]: Good catch - this is a potential issue:  https://github.com/hpcc-systems/Visualization/issues/3051

File: packages/marshaller/src/ddl2/model/element.ts (hunk 3 above)
Comment (GordonSmith) [reply packages/marshaller/src/ddl2/model/element.ts]: Here is the scenario - create a dashboard with 4 widgets and then delete the first 2 and save it.

Now when you open it the max ID is e_4 - but there are only "2" widgets so the internal counter is now at 2

At some point there is the potential that there will be a collision when the user adds another 2 widgets...

**Pull request merged**